### PR TITLE
Using security scopes in m4

### PR DESCRIPTION
**Pull request by qiaozha**
to resolve issues 

- https://github.com/Azure/autorest.typescript/issues/1322
- https://github.com/Azure/autorest.typescript/issues/1356
- https://github.com/Azure/autorest.typescript/issues/1385

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -70,14 +70,15 @@ modelerfour:
   group-parameters: true
   # Enable older inconsistent behavior that an enum with a single value would become a constant by default.
   seal-single-value-enum-by-default: true
+  lenient-model-deduplication: true
```
Comment (joheredi) [inline README.md:73]: I don't think we should enable this by default. This should be enabled in a case by case since it works around buggy swaggers and slows down the generation dramatically

File: src/autorestSession.ts [hunk 2]
```diff
@@ -12,9 +12,9 @@ export interface AutorestOptions {
   restLevelClient?: boolean;
   rlcShortcut?: boolean;
   azureArm?: boolean;
-  addCredentials?: boolean;
```
Comment (joheredi) [inline src/autorestSession.ts:15]: I think we should keep these settings available for backward compatibility. If a swagger has security scopes and also these passed in the config, these should win over the swagger's

File: src/generators/clientFileGenerator.ts [hunk 3]
```diff
@@ -559,10 +560,11 @@ function writeDefaultOptions(
   hasLro: boolean,
   clientDetails: ClientDetails
 ) {
-  const { useCoreV2, credentialScopes, packageDetails } = getAutorestOptions();
+  const { useCoreV2, packageDetails } = getAutorestOptions();
+  const { credentialScopes } = getSecurityInfoFromModel(clientDetails.security);
 
   const credentialScopesValues = getCredentialScopesValue(credentialScopes);
-  const addScopes = credentialScopes
+  const addScopes = credentialScopes && credentialScopes.length > 0
```
Comment (joheredi) [inline src/generators/clientFileGenerator.ts:567]: Here we can check if we have addCredentials and the old settings available in `getAutorestOptions` if so, use them otherwise use the settings from `getSecurityInfoFromModel`

File: src/main.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ import {
 export async function processRequest(host: AutorestExtensionHost) {
   await initializeSession(host);
   const session = getSession();
-  const { restLevelClient } = getAutorestOptions();
+  const { restLevelClient, azureArm } = getAutorestOptions();
```
Comment (joheredi) [inline src/main.ts:19]: Is this being used?

File: src/restLevelClient/generateClient.ts [hunk 5]
```diff
@@ -153,7 +165,10 @@ function getClientFactoryBody(
     declarations: [{ name: "baseUrl", initializer: baseUrl }]
   };
 
-  const { credentialScopes, credentialKeyHeaderName } = getAutorestOptions();
+  const {
```
Comment (joheredi) [inline src/restLevelClient/generateClient.ts:168]: Here as well check if getAutorestOptions have credentialScopes or credentialKeyHeaderName and use it, otherwise try to get security info from `getSecurityInfoFromModel`

File: src/transforms/parameterTransforms.ts [hunk 6]
```diff
@@ -104,7 +105,7 @@ export function transformParameters(
   options: ClientOptions
 ): ParameterDetails[] {
   let parameters: ParameterDetails[] = [];
-  const { addCredentials } = getAutorestOptions();
+  const { addCredentials } = getSecurityInfoFromModel(codeModel.security);
```
Comment (joheredi) [inline src/transforms/parameterTransforms.ts:108]: same here, try to get it from AutorestOptions and fallback to getSecurityInfoFromModel

File: src/utils/autorestOptions.ts [hunk 7]
```diff
@@ -167,24 +167,17 @@ async function getOutputPath(
   return (await host.getValue("output-folder")) || undefined;
 }
 
-async function getKeyCredentialHeaderName(
+async function getSecurityHeaderName(
   host: AutorestExtensionHost
 ): Promise<string | undefined> {
-  return (await host.getValue("credential-key-header-name")) || undefined;
+  return (await host.getValue("security-header-name")) || undefined;
```
Comment (joheredi) [inline src/utils/autorestOptions.ts:173]: Same here

File: src/utils/autorestOptions.ts [hunk 8]
```diff
@@ -256,38 +249,14 @@ async function getPackageDetails(
   };
 }
 
-export async function getCredentialScopes(
+export async function getSecurityScopes(
   host: AutorestExtensionHost
 ): Promise<string[] | undefined> {
-  const addCredentials = await host.getValue("add-credentials");
```
Comment (joheredi) [inline src/utils/autorestOptions.ts:262]: We should keep supporting the old way

File: test/integration/generated/constantParam/src/generatedClient.ts [hunk 9]
```diff
@@ -23,11 +24,19 @@ export class GeneratedClient extends coreClient.ServiceClient {
...
   /**
    * Initializes a new instance of the GeneratedClient class.
+   * @param credentials Subscription credentials which uniquely identify client subscription.
    * @param endpoint Supported Cognitive Services endpoints (protocol and hostname, for example:
    *                 https://westus.api.cognitive.microsoft.com).
    * @param options The parameter options
    */
-  constructor(endpoint: string, options?: GeneratedClientOptionalParams) {
+  constructor(
+    credentials: coreAuth.TokenCredential,
```
Comment (joheredi) [inline test/integration/generated/constantParam/src/generatedClient.ts:33]: Why are we now requiring credentials here? These tests don't take credentials

File: test/integration/generated/corecompattest/src/petStore.ts [hunk 10]
```diff
@@ -77,49 +76,15 @@ export class PetStore extends coreHttpCompat.ExtendedServiceClient {
 
   /**
    * Fake endpoint to test byte array in body parameter for adding a new pet to the store
-   * @param contentType Upload file type
    * @param options The options parameters.
    */
   addPetUsingByteArray(
-    contentType: "application/json",
```
Comment (joheredi) [inline test/integration/generated/corecompattest/src/petStore.ts:84]: Where is this change comming from?

File: test/integration/generated/datafactory/src/dataFactoryClient.ts [hunk 11]
```diff
@@ -52,13 +53,18 @@ export class DataFactoryClient extends coreClient.ServiceClient {
 
   /**
    * Initializes a new instance of the DataFactoryClient class.
+   * @param credentials Subscription credentials which uniquely identify client subscription.
    * @param subscriptionId The subscription identifier.
    * @param options The parameter options
    */
   constructor(
+    credentials: coreAuth.TokenCredential,
```
Comment (joheredi) [inline test/integration/generated/datafactory/src/dataFactoryClient.ts:61]: Looks like some default setting changed and now everything is getting a credential?

File: test/integration/generated/datafactory/src/models/mappers.ts [hunk 12]
```diff
@@ -1661,7 +1659,6 @@ export const LinkedServiceListResponse: coreClient.CompositeMapper = {
 };
 
 export const LinkedService: coreClient.CompositeMapper = {
-  serializedName: "LinkedService",
```
Comment (joheredi) [inline test/integration/generated/datafactory/src/models/mappers.ts:1664]: What's going on with this? Why is serialized-name removed?

File: test/integration/generated/mediaTypes/src/mediaTypesClient.ts [hunk 13]
```diff
@@ -204,135 +201,32 @@ export class MediaTypesClient extends coreClient.ServiceClient {
...
    * {'hello': world'} with content type 'application/json' and a byte string for
    * 'application/octet-stream'.
-   * @param contentType Upload file type
    * @param message The payload body.
    * @param options The options parameters.
    */
   binaryBodyWithThreeContentTypes(
-    contentType: ContentType1,
     message: coreRestPipeline.RequestBodyType,
-    options?: BinaryBodyWithThreeContentTypes$binaryOptionalParams
```
Comment (joheredi) [inline test/integration/generated/mediaTypes/src/mediaTypesClient.ts:214]: Looks like something around media-types broke

**Review by joheredi — CHANGES_REQUESTED** [review]
There are a few changes in the generated code that worry me. It looks like 1) now we are always requiring credentials and 2) something seems to have broken in the media-type overloads.

Another comment I have throughout the PR is to keep supporting the old configs `add-credentials` and the others fro backward compatibility.

File: test/integration/generated/polymorphicSkipNormalize/src/generated/mediaServicesClient.ts [hunk 14]
```diff
@@ -1,18 +1,28 @@
...
 
 export class MediaServicesClient extends coreClient.ServiceClient {
   /**
    * Initializes a new instance of the MediaServicesClient class.
+   * @param credentials Subscription credentials which uniquely identify client subscription.
    * @param options The parameter options
    */
-  constructor(options?: MediaServicesClientOptionalParams) {
+  constructor(
+    credentials: coreAuth.TokenCredential,
```
Comment (qiaozha) [inline test/integration/generated/polymorphicSkipNormalize/src/generated/mediaServicesClient.ts:12]: I found this change is because the swagger has such definition https://github.com/Azure/azure-rest-api-specs/blob/a57e89f8c3e1215201694cc0d6e9a51842573df0/specification/mediaservices/data-plane/LiveVideoAnalytics.Edge/preview/2.0.0/LiveVideoAnalyticsSdkDefinitions.json#L17-L23

Comment (qiaozha) [reply test/integration/generated/polymorphicSkipNormalize/src/generated/mediaServicesClient.ts:12]: even if we have set security: undefined, the m4 seems still generate the security.

File: test/integration/generated/mediaTypesWithTracing/src/models/index.ts [hunk 15]
```diff
@@ -99,12 +105,11 @@ export type BinaryBodyWithThreeContentTypesResponse = {
...
-export interface PutTextAndJsonBody$textOptionalParams
-  extends coreClient.OperationOptions {}
-
-/** Optional parameters. */
-export interface PutTextAndJsonBody$jsonOptionalParams
-  extends coreClient.OperationOptions {}
+export interface PutTextAndJsonBodyOptionalParams
+  extends coreClient.OperationOptions {
+  /** Upload file type */
+  contentType?: ContentType1AutoGenerated2;
```
Comment (qiaozha) [inline test/integration/generated/mediaTypesWithTracing/src/models/index.ts:111]: looks like the contentType parameter becomes optional parameter

File: package.json [hunk 16]
```diff
@@ -46,10 +46,10 @@
         "LICENSE"
     ],
     "dependencies": {
-        "@autorest/codemodel": "^4.17.0",
+        "@autorest/codemodel": "^4.18.2",
         "@autorest/extension-base": "^3.4.1",
         "@azure-rest/core-client": "1.0.0-alpha.20220112.1",
-        "@autorest/testmodeler": "^1.0.4",
+        "@autorest/testmodeler": "^2.2.4",
```
Comment (joheredi) [inline package.json:52]: Can we do changes that involve testmodeler in a separate PR?

Comment (qiaozha) [reply package.json:52]: actually if we want to use the security scopes in code model, we need to upgrade the modelerfour version and the older test modeler doesn't compatible with the latest modelerfour. so we need to upgrade them both.
About the testmodeler CG issue, changlong is working on it.

File: test/integration/generated/polymorphicSkipNormalize/src/generated/mediaServicesClient.ts (hunk 14 above)
Comment (joheredi) [reply test/integration/generated/polymorphicSkipNormalize/src/generated/mediaServicesClient.ts]: Due to https://github.com/Azure/autorest.typescript/issues/1373 we currently don't have the testmodeler dependency so we should be able to have it in a separate PR

File: package.json (hunk 16 above)
Comment (joheredi) [reply package.json:52]: Due to https://github.com/Azure/autorest.typescript/issues/1373 we currently don't have the testmodeler dependency so we should be able to have it in a separate PR

File: test/integration/generated/constantParam/src/generatedClient.ts [hunk 17]
```diff
@@ -70,15 +70,17 @@ export class GeneratedClient extends coreClient.ServiceClient {
    * indicate 100% certainty that the identified language is true. See the <a
    * href="https://aka.ms/talangs">Supported languages in Text Analytics API</a> for the list of enabled
    * languages.
+   * @param contentType Body Parameter content-type
    * @param input Collection of documents to analyze for language endpoint.
    * @param options The options parameters.
    */
   languages(
+    contentType: "application/json",
```
Comment (joheredi) [inline test/integration/generated/constantParam/src/generatedClient.ts:78]: If there is one content-type defined in the swagger, can we make contentType optional, put it in the options bag?

**Comment by joheredi** [comment]
This is looking good! I think we are almost there. Left a comment about handling the case where there is only a single content-type

**Comment by qiaozha** [comment]
close this one in favor of #1396
